src/node/oscillator.rs

```diff
@@ -406,242 +225,77 @@ impl OscillatorNode {
-    fn change_type(&self, type_: OscillatorType) {
-        self.type_.store(type_ as u32, Ordering::SeqCst);
-    }
-
-    /// set the oscillator type to custom and generate
-    /// a perdioc waveform following the `PeriodicWave` characteristics
+    /// Calling this sets the oscillator type to `custom`, once set to `custom`
+    /// the oscillator cannot be reverted back to a standard waveform.
     pub fn set_periodic_wave(&self, periodic_wave: PeriodicWave) {
-        // The oscillator type is set to custom following the spec
-        self.change_type(OscillatorType::Custom);
-
-        let PeriodicWave {
-            real,
-            imag,
-            disable_normalization,
-        } = periodic_wave;
-
-        let mut cplxs = Vec::with_capacity(real.len());
-        let mut norms = Vec::with_capacity(real.len());
-        let mut phases = Vec::with_capacity(real.len());
-        let mut incr_phases = Vec::with_capacity(real.len());
-        let mut interpol_ratios = Vec::with_capacity(real.len());
-
-        // update cplxs
-        for cplx in real.into_iter().zip(imag) {
-            cplxs.push(cplx);
-        }
-
-        let computed_freq = self.computed_freq();
-
-        for (idx, (real, img)) in cplxs.iter().enumerate() {
-            // update norms
-            norms.push((f32::powi(*real, 2_i32) + f32::powi(*img, 2_i32)).sqrt());
-
-            // update phases
-            let phase = f32::atan2(*img, *real);
-            if phase < 0. {
-                phases.push(2.0_f32.mul_add(PI, phase) * (TABLE_LENGTH_F32 / (2.0 * PI)));
-            } else {
-                phases.push(phase * (TABLE_LENGTH_F32 / (2.0 * PI)));
-            }
-
-            // update incr_phases
-            // 0 through max value 8192 casts without loss of precision
-            #[allow(clippy::cast_precision_loss)]
-            incr_phases.push(
-                TABLE_LENGTH_F32 * idx as f32 * (computed_freq / self.context().sample_rate()),
-            );
-        }
-
-        // update interpol_ratios
-        for incr_phase in &incr_phases {
-            interpol_ratios.push((incr_phase - incr_phase.round()).abs());
-        }
-
-        // generate the wavetable following periodic wave characteristics
-        let dyn_wavetable =
-            Self::generate_wavetable(&norms, &mut phases, &incr_phases, &interpol_ratios);
-
-        // update norm_factor
-        let norm_factor = Self::norm_factor(&dyn_wavetable);
+        self.type_
+            .store(OscillatorType::Custom as u32, Ordering::SeqCst);
 
         self.sender
-            .send(OscMsg::PeriodicWaveMsg {
-                computed_freq,
-                dyn_wavetable,
-                norm_factor,
-                disable_normalization,
-            })
+            .send(OscillatorMessage(periodic_wave))
             .expect("Sending periodic wave to the node renderer failed");
     }
-
-    /// Generate the wavetable
```

Comment:
Great to see you have decoupled the OscNode from the PeriodicWave stuff. Much more succinct this way

Reply:
Yup, but I have to say that this implementation might be a bit less efficient than the previous one, I don't really know. But there was several issues with the previous implementation (the `real` and `imaginary` components were somehow inverted, and there was a small offset in the phase), so I though it would be better to just stick to the mathematical formula, write the tests against it, so a next iteration for perf improvement (e.g. iFFT or table lookup) would be easier and accurate.

Reply:
Correctness is most important obviously. We can always measure and optimize later